### PR TITLE
#1901 refactor review change request logic

### LINKED ISSUE #1901: Auto Implementing CRs - Add Proposed Changes Logic to Review CR Endpoint
### Description

Fair warning, this one is gonna be a bitch

When we review a standard change request (scope change request) we need to check if it has any wbs proposed changes (and by extension work package or project proposed changes) on it. If it does, you should go through and make all the necessary changes on the work package or project respectively using the proposed changes in addition to the existing logic. 

### Acceptance Criteria

- [ ] Apply all changes from the wbs and work package or project proposed changes on the scope cr if it has them
- [ ] ensure that the wbs element has no other open, unreviewed CRs (wanna avoid merge conflicts)

### Proposed Solution

Important stuff to note: if a crID associated with a project has work package proposed changes, then it is creating a new work package. Edits on projects will have 'newProject' set to false, and creates will have it set to true

### Mocks

epic #1892

### PR DESCRIPTION
**Pull request by bderbs30**
## Changes

refactored reviewChangeRequest scopeCR logic for work and project proposed changes

## Notes

For the first test case it doesn't seem to be creating the wp lead and manager correctly so that needs to be fixed possibly. It doesn't seem to be done in the create service at all. It seems you can only set this in the edit part.

## Test Cases

1. If a crId has an associated project and also has workpackage proposed changes then we should be creating a new workpackage with the workpackage proposed changes. Thus with the above condition we need to check that a new workpackage was created with the correct proposed changes
image
image
image
image

2.. If the CR doesn't have an associated project but has workpackage proposed changes then we should be editing that workpackage with the workpackage proposed changes. Thus with the above condition we need to check that the workpackage was edited with the correct proposed changes

3.. If a CR doesn't have workpackage proposed changes but has project proposed changes and is a newProject(true) then we should be creating a new project with the project proposed changes. Thus with the above condition we need to check that a new project was created with the correct project proposed changes.

4. If the newProject is FALSE then we should be editing a project with the project proposed changes. Thus with the above condition we need to check that the project was edited with the correct project proposed changes
image
**This kinda shows it I actually need proper tests but these projects had projectproposedchanges and then were reviewed (in the seed file) and thus changed to the proposed changes**

## Screenshots

_If you did any manual testing (e.g., with Postman), put screenshots of the http request and before and after of the db_

_If none of this applies, you can delete this section_

## To Do

- [ ] tests
- [x] editing project logic
- [x] editing workpackage logic

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #1901

### CODE REVIEW

File: src/backend/src/services/change-requests.services.ts [hunk 1]
```diff
@@ -104,8 +110,10 @@ export default class ChangeRequestsService {
...
     // verify that the user is not reviewing their own change request
-    if (reviewer.userId === foundCR.submitterId) throw new AccessDeniedException();
-
+    if (reviewer.userId === foundCR.submitterId)
+      throw new AccessDeniedException(
+        `User ${reviewer.userId} is not allowed to review their own change request submitted by User ${foundCR.submitterId}`
+      );
     // If approving a scope CR
     if (foundCR.scopeChangeRequest && accepted) {
       // ensure a proposed solution is being approved and exists
```
Comment (RChandler234) [inline src/backend/src/services/change-requests.services.ts:119]: change this, if no wbs proposed changes and no psid, then error
if no proposedsolutions, but proposedchanges: do your new stuff
if no proposedchanges, but proposedsolutions: do old stuff

File: src/backend/src/services/change-requests.services.ts [hunk 2]
```diff
@@ -202,6 +210,108 @@ export default class ChangeRequestsService {
...
+            features: projProposedChanges.features.map(({ id, detail }) => ({ id, detail })),
+            otherConstrains: projProposedChanges.otherConstrains.map(({ id, detail }) => ({ id, detail })),
+            projectLeadId: projProposedChanges.projectLead?.userId ?? null,
+            projectManagerId: projProposedChanges.projectManager?.userId ?? null
+          };
+        };
+
+        // if you have project proposed changes
+        if (projectProposedChanges) {
+          const projProposedChanges = transformProjectData(projectProposedChangesTransformer(wbsProposedChanges));
```
Comment (bderbs30) [inline src/backend/src/services/change-requests.services.ts:274]: Change how we transform data this is kinda funky

**Review by RChandler234 — CHANGES_REQUESTED** [review]

File: src/backend/src/services/change-requests.services.ts [hunk 3]
```diff
@@ -202,6 +210,108 @@ export default class ChangeRequestsService {
...
+
+        await prisma.change_Request.update({
+          where: { crId },
+          data: {
+            accepted: true,
+            dateReviewed: new Date()
+          }
+        });
+
+        const associatedProjectCR = foundCR.wbsElement.project;
```
Comment (bderbs30) [inline src/backend/src/services/change-requests.services.ts:232]: if the cr has a associated wp and wpchanges then its editing the wp

File: src/backend/src/utils/utils.ts [hunk 4]
```diff
@@ -23,3 +23,9 @@ export const validateInputs = (req: Request, res: Response, next: Function): Res
   }
   next();
 };
+
+export const transformDate = (date: Date) => {
+  const month = date.getMonth() + 1 < 10 ? `0${date.getMonth() + 1}` : (date.getMonth() + 1).toString();
+  const day = date.getDate() < 10 ? `0${date.getDate()}` : date.getDate().toString();
+  return `${date.getFullYear().toString()}-${month}-${day}`;
+};
```
Comment (RChandler234) [inline src/backend/src/utils/utils.ts:31]: ok don't know how I didn't catch this before, but we should not have a file called utils.ts in a folder called utils, just kinda redundant. Pls move transform date out of here into another file (we should probably have a datetime utils file in the backend as well), I'll make another ticket to remove this file since it's gonna cause a lot of import changes, dw hold this PR up

File: src/backend/src/services/change-requests.services.ts [hunk 5]
```diff
@@ -104,108 +114,215 @@ export default class ChangeRequestsService {
...
-          reviewer.userId,
-          foundCR.wbsElementId
-        );
-        await prisma.project.update({
-          where: { projectId: foundCR.wbsElement.project.projectId },
-          data: {
-            budget: newBudget
-          }
-        });
+      if (!foundCR.scopeChangeRequest.wbsProposedChanges && !psId) {
```
Comment (RChandler234) [inline src/backend/src/services/change-requests.services.ts:123]: you're kind of just the straw that broke the camel's back here, but this is way too much code in one method. To start off, can you make all these higher level if statements into variables above this with good names, so you can tell exactly what's happening in the if statement e.g. if(isApprovingScopeChangeRequest)

Comment (RChandler234) [reply src/backend/src/services/change-requests.services.ts:123]: secondly, each if statement could probably be it's own helper method. Tbh we could just make them methods on this service class as well. If ends up looking gross, feel free to move them to change request utils but I think most of these things are high level enough that they warrant a service method, just make sure they're each documented well in a comment

File: src/backend/src/utils/change-requests.utils.ts [hunk 6]
```diff
@@ -182,3 +182,17 @@ export const validateProposedChangesFields = async (
...
+
+/**
+ * Determines if there are other open unreviewed change requests for this wbs element
+ * @param wbsElemId the wbs element id to find CRs with
+ * @throws if the WBS element has open unreviewed change requests
+ *
+ */
+export const validateUnreviewedOpenCRs = async (wbsElemId: number) => {
+  const openCRs = await prisma.change_Request.findMany({
+    where: { wbsElementId: wbsElemId, accepted: false }
```
Comment (RChandler234) [inline src/backend/src/utils/change-requests.utils.ts:194]: oh I see this is so we don't have merge conflicts ever. Add a comment to where this is called to specify that's why we need to enforce this invariant, I think this should be looking for any change request where dateReviewed is undefined instead. We don't want this error to throw if there are any denied change requests on the same wbs element

**Review by RChandler234 — CHANGES_REQUESTED** [review]

File: src/backend/src/services/change-requests.services.ts [hunk 7]
```diff
@@ -104,204 +151,20 @@ export default class ChangeRequestsService {
...
-          workPackage: { update: { startDate: actCr.startDate } },
-          status: WBS_Element_Status.ACTIVE
-        }
-      });
+      await this.reviewScopeChangeRequest(foundCR, reviewer, psId);
+    } else if (accepted && foundCR.type === CR_Type.STAGE_GATE) {
+      // stage gate cr
+      await this.reviewStageGateChangeRequest(foundCR, reviewer);
+    } else if (foundCR.type === CR_Type.ACTIVATION && foundCR.activationChangeRequest && accepted) {
+      // if it's an activation cr and being accepted, we can do some stuff to the associated work package
```
Comment (RChandler234) [inline src/backend/src/services/change-requests.services.ts:166]: move these comments to right above their corresponding if statement, and clean them up, e.g. "Stage Gate Change Request"

File: src/backend/src/services/change-requests.services.ts [hunk 8]
```diff
@@ -335,6 +198,120 @@ export default class ChangeRequestsService {
...
+          associatedProject,
+          reviewer,
+          foundCR.crId,
+          foundCR.wbsElement.carNumber
+        );
+      }
+    }
+  }
+
+  static async reviewActivationChangeRequest(foundCR: any, reviewer: User): Promise<void> {
```
Comment (RChandler234) [inline src/backend/src/services/change-requests.services.ts:251]: nice work with these, make sure to add a comment for each of them describing the inputs/ what they're doing

File: src/backend/src/services/change-requests.services.ts [hunk 9]
```diff
@@ -335,6 +198,120 @@ export default class ChangeRequestsService {
...
+      // if there isn't wbs changes or proposed solutions
+      throw new HttpException(400, 'No proposed solution or proposed changes for scope change request');
+    } else if (psId && !foundCR.scopeChangeRequest.wbsProposedChanges) {
+      // if there is only a proposed solution and no wbs changes
+
+      // reviews a proposed solution applying certain changes based on the content of the proposed solution
+      await reviewProposedSolution(psId, foundCR, reviewer);
+    } else if (foundCR.scopeChangeRequest.wbsProposedChanges && !psId) {
+      await validateNoUnreviewedOpenCRs(foundCR.wbsElementId);
+      // we don't want to have merge conflict on the wbs element thus we check if there are unreviewed or open CRs on the wbs element
```
Comment (RChandler234) [inline src/backend/src/services/change-requests.services.ts:212]: comments should always be above the line that they're describing, make sure to make  that change everywhere in this PR

File: src/backend/src/services/change-requests.services.ts [hunk 10]
```diff
@@ -66,6 +79,40 @@ export default class ChangeRequestsService {
     return changeRequests.map(changeRequestTransformer);
   }
 
+  static async reviewStageGateChangeRequest(foundCR: any, reviewer: User): Promise<void> {
```
Comment (RChandler234) [inline src/backend/src/services/change-requests.services.ts:82]: maybe put this down with the other two just for organization purposes

File: src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/TaskList/TaskListTabPanel.tsx [hunk 11]
```diff
@@ -1,198 +1,199 @@
...
-          teams={teams}
-          hasEditPermissions={editTaskPermissions(selectedTask!)}
-        />
-      )}
-    </Box>
-  );
-};
-
-export default TaskListTabPanel;
+/*
```
Comment (RChandler234) [inline src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/TaskList/TaskListTabPanel.tsx:1]: this looks like an unintentional edit, pls revert the changes to this file, unless you were changing the imports or smthn?

File: src/frontend/src/utils/change-requests.utils.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import { ChangeRequest, StandardChangeRequest } from 'shared';
+
+export const canReviewChangeRequest = (cr: ChangeRequest, selected: number) => {
```
Comment (RChandler234) [inline src/frontend/src/utils/change-requests.utils.ts:3]: this is a pretty weird helper method, I would just keep it in the file where it's being used

File: src/backend/src/utils/change-requests.utils.ts [hunk 13]
```diff
@@ -182,3 +186,226 @@ export const validateProposedChangesFields = async (
...
+ */
+export const validateNoUnreviewedOpenCRs = async (wbsElemId: number) => {
+  const openCRs = await prisma.change_Request.findMany({
+    where: { wbsElementId: wbsElemId, dateReviewed: null }
+  });
+  if (openCRs.length > 1)
+    throw new HttpException(400, 'There are other open unreviewed change requests for this WBS element');
+};
+
+// will replace any type in service
```
Comment (RChandler234) [inline src/backend/src/utils/change-requests.utils.ts:204]: pls do and remove comment

File: src/backend/src/utils/change-requests.utils.ts [hunk 14]
```diff
@@ -182,3 +186,226 @@ export const validateProposedChangesFields = async (
...
+// export type ChangeRequestWithFK = Change_Request & {
+//   wbsElement:
+
+// }
+
+export const applyProjectProposedChanges = async (
```
Comment (RChandler234) [inline src/backend/src/utils/change-requests.utils.ts:210]: can you add a nice comment to each of these utils method

**Review by RChandler234 — CHANGES_REQUESTED** [review]
This is looking a lot better, can u lmk in a comment on the PR if there is any functionality that you've tested that isn't working

File: src/frontend/src/utils/change-requests.utils.ts (hunk 12 above)
Comment (bderbs30) [reply src/frontend/src/utils/change-requests.utils.ts:3]: I might've misunderstood but you said to make this a helper method

Comment (RChandler234) [reply src/frontend/src/utils/change-requests.utils.ts:3]: ohhh I see, sorry should've been more specific. The second half, not the selected stuff, should be a helper method and you can name is "hasProposedChanges" or smthn

**Comment by RChandler234** [comment]
confirmed all existing cr functionality (not slack stuff tho, seed data being rly annoying in my local)

**Review by walker-sean — DISMISSED** [review]
